releasewarrior/helpers.py

```diff
@@ -206,12 +206,13 @@ def validate(release, logger, config, must_exist=False, must_exist_in=None):
 def validate_data_repo_updated(logger, config):
     repo = Repo(config['releasewarrior_data_repo'])
     if repo.is_dirty():
-        logger.warning("release data repo dirty")
+        logger.fatal("release data repo dirty. Aborting...")
+        return False
     upstream = find_upstream_repo(repo, logger, config)
     # TODO - we should allow csets to exist locally that are not on remote.
     logger.info("ensuring releasewarrior repo is up to date and in sync with {}".format(upstream))
-    logger.debug('fetching new csets from {}/master'.format(upstream))
-    upstream.fetch()
+    logger.debug('pulling new csets from {}/master'.format(upstream))
+    upstream.pull(rebase=True)
     commits_behind = list(repo.iter_commits('master..{}/master'.format(upstream)))
```

Comment:
perhaps, as a way to address the issue raised by @tomprince but also make things more convenient, we should instead be looking for `commits_ahead`. e.g. `'{}/master..master'.format(upstream))`.  This means we could:

```python
upstream.fetch()
if repo.is_dirty() or commits_ahead: 
    logger.fatal()
else:
    upstream.merge() # or w/e the GItPython method name is here.
```

This way, we automatically fetch and merge in the most common case. And in the rare case where we have local changes or commits that are not upstream, it's left to the user to rebase, fast-forward, w/e?

I think

Reply:
Thank you for the idea! To clarify, what is `w/e` in this context?

So we have 4 different scenarios depending on the ratio risk/convenience we want to have:

## What we've had so far in rw2

`if commits behind then fail`.


## @lundjordan's proposal

`if just commits behind then merge, but if commits ahead then fail`. This makes rw2 fail at the second command. Eg:
```sh
$ release track firefox 99.0b3
[...]
INFO: committing changes with message: firefox 99.0b3 started tracking upcoming release.
$ release status
CRITICAL: [Error from logger.fatal() in snippet from the previous command]
```
Therefore, every time we make a change, we must push it.

## @tomprince's `--ff` proposal

`if (commits behind or commits ahead but none behind) then merge --ff, otherwise fail`. This can be done via `upstream.pull(ff_only=True)`.

> :warning: I realized GitPython uses `~/.gitconfig`. Mine was configured to rebase on pull. So `ff_only` was ignored. Once I got rid of it, I could test this case out.

One case where it's inconvenient is:
> a. Squirrel A makes a local change on Release X but doesn't push.
> b. Squirrel B makes a local change on Release Y *and* pushes.
> c. Squirrel A wants to make a second local change on Release X => fails, even though there no changes in Release X

## The rebase proposal

`if commits behind and ahead, try to rebase. Otherwise just update the latest`. If lines are changed the same way (like a task being resolved by both Squirrel A and B), no error comes up. One case it fails is: when Squirrel A and B change the same line with different values (like a taskgraphId) and none of them had pushed anything. For instance, here what could be the output of such an error:

```sh
$ release status
INFO: ensuring releasewarrior repo is up to date and in sync with upstream
CRITICAL: Could not rebase the repo on top of upstream/master. Please run `git rebase --abort` then `git pull --rebase` then resolve the conflicts manually
```

> :warning: Unlike `ff_only=True`, `rebase=True` ignores `ff = only` in my `~/.gitconfig`. I don't know why.

## To sum up

| Solution | Biggest annoyance | Probability |
|:-------------:|-------------|:-----:|
| 1 | Fails whenever a commit in upstream is not present locally | High | 
| 2 | Fails whenever a local commit is not present upstream | Highest | 
| 3 | Fails whenever changes are made in parallel, even though they don't impact the same files | Medium | 
| 4 | Fails whenever the same parts of the same files are modified with different values | Low | 

As a conclusion, I think merge conflicts (the ones that can't be resolved automatically by git) are rare compared to what the other methods impose to Squirrel A and B when they work in parallel. I personally prefer a rare breakage once in while than juggling between the 2 repos regularly. However, I understand not everybody is ready to take this risk. I'm fine using one of the aforementioned solution.

What do you guys think?

Reply:
> to clarify, what is w/e in this context?

whatever. As in I was using pseudo code and meant whatever the actual syntax and naming is

> Therefore, every time we make a change, we must push it.

Thanks for reminding me why I didn't end up going with this approach in my first draft :)

I'm happy giving either (3) or (4) an approach. If it's annoying, we can always revert. I think either will be better than our current situation (1). Since you have (4) done in this PR, and I'd hate to spend more time on rw than needed, my vote is that.

Reply:
@tomprince ?

Reply:
I know for myself, I don't want auto-merge or rebase on random commands, so if (4) is chosen. It sounds like different people have different expectations, so perhaps we need to make this configurable.

What I don't want is for the rw-data repository to be left in the middle of a rebase when running a command that isn't about interacting with the remote repository.